### DISCUSSION TITLE
Global data storage

### DISCUSSION THREAD

**Discussion by iBobik** [nuxt/nuxt#16479] [Questions]
Hi, how to properly fetch data once but use everywhere?

Common use-case is menu tree. If I fetch that data on the menu component then it will be fetched for every route. I saw some examples using Vuex and serverInit hook, but it is overkill if state management is not needed anywhere else and Nuxt 3 does not support it.

Or if we can fetch all pages in one cheap API call why not do it. Usual solution is routes in nuxt.config.yml, but this can not use bootstrapped context and does not work in dev.

**Comment by procrates** [comment]
Try using shared state with composables: [https://v3.nuxtjs.org/guide/features/state-management#shared-state](https://v3.nuxtjs.org/guide/features/state-management#shared-state).
Fetch the data and populate it in the shared state. 

Then access the navigation anywhere with:
`const navigation = useNavigation()`

**Reply by iBobik** [reply]
I see, but is there a way how to have that fetch code somewhere else than in every component? Like in Nuxt2 stores folder?

**Reply by procrates** [reply]
Essentially this is a store. Just like in nuxt 2 where you had to use vuex and computed to get the state you just need to add that composable state where you need it. If you add it to app.vue or layout it should be accessible everywhere if I am not totally wrong.

Another solution would be to use pinia which I my preferred option which is basically vuex without mutations, and has alot better syntax(in my opinion).

Both pinia and shared state can set local storage or cookies to preserve state between refresh.

**Reply by procrates** [reply]
Oh I read your reply wrong.

Make a composable to fetch the data and run it in app.vue or layout.

If you don't want it to run every time just do a if check to see if the shared state is populated else run the fetch.

Your menu fetch is now in one place and is running globally and will not refetch until you refresh or delete it.

**Reply by eskiesirius** [reply]
how to populate shared state with a new value?

**Reply by chrisnoden** [reply]
It's not clear if this creates a state that is shared across many visitors to your site or if the state will be generated for each visitor. I am after a solution where I can fetch data into the server that is then available to all visitors without having to fetch it from the remote API when each new visitor first lands on the site.

**Comment by idflood** [comment]
I like the idea of @procrates but can't make it to work. It essentially load the data on every page instead of reusing the state. 

Am I missing something?

```javascript
// in composable/useGlobalSettings.ts

export const useGlobalSettings = async () => {
  const GqlInstance = useGql()

  const settings = useState('settings', () => null)
  if (!settings.value) {
    console.log("loading global settings") // this is called everytime but should only be displayed once
    settings.value = await useAsyncData('globalSettings', () =>  GqlInstance('globalSettings', {}))
  }

  return {
    settings
  }
}

```

**Comment by artem-kostylev** [comment]
@idflood  I think it should be like this

```ts
import { useAsyncData, useNuxtData } from '#imports';
import type { AsyncDataOptions } from '#app';

export const useCachedAsyncData = <T>(key: string, cb: () => Promise<T>, options?: AsyncDataOptions<T, any, any>) => {
    return useAsyncData<T>(key, () => {
        const { data } = useNuxtData(key);

        return data.value ?? cb();
    }, options);
}
```

**Reply by idflood** [reply]
@artem-kostylev thanks, I tested it but it still loads data for every route. I also tried to log a `useNuxtData` after the call to `useCachedAsyncData`, it's correctly set but seems to get lost after each pages.

**Reply by idflood** [reply]
I've created a minimal [test example](https://github.com/idflood/nuxt3-static-test/tree/cached-data-test) on (cached-data-test branch) in case this was a conflict with something else in my projects.

After adding a `console.log` in the `useCachedAsyncData` function above when data.value is null I can see that it still fetch the data on every routes when running `npm nuxi generate`.

![Screenshot 2023-01-28 at 11 42 51](https://user-images.githubusercontent.com/197418/215262313-3c5c8d91-322b-451a-b52f-4fd63cf7fa76.png)


edit: Also if I add these 2 lines just before the end of that function I see that the `nuxt.payload` and `nuxt.static` objects are not persisted across routes.

```javascript
const nuxt = useNuxtApp();
console.log(nuxt.static.data, nuxt.payload.data);
```
